dbdimp.c

```diff
@@ -5917,78 +5939,77 @@ int mariadb_db_async_ready(SV* h)
 
 static bool parse_number(char *string, STRLEN len, char **end)
 {
-    int seen_neg = 0;
     bool seen_dec = FALSE;
     bool seen_e = FALSE;
-    bool seen_plus = FALSE;
-    char *cp;
-
-    if (len <= 0) {
-        len= strlen(string);
-    }
+    bool seen_digit = FALSE;
```

Comment:
It looks horribly complicated. What about something like:
```
static bool parse_number(char *string, STRLEN len, char **end)
{
    char *cp = string;
    bool number_found = FALSE;

    /* Skip leading whitespace */
    while (*cp == ' ')
      cp++;

    /* Optional '+' or '-' */
    if (*cp == '-' || *cp == '+')
      cp++;

    /* Number before '.' */
    while (*cp >= '0' && *cp <= '9')
    {
      cp++;
      number_found = TRUE;
    }

    /* Optional '.' */
    if (*cp == '.')
    {
      cp++;

      /* Number after '.' */
      while (*cp >= '0' && *cp <= '9')
      {
        cp++;
        number_found = TRUE;
      }
    }

    /* No number found - error */
    if (!number_found)
        return FALSE;

    /* Optional exponent */
    if (*cp == 'e' || *cp == 'E')
    {
      cp++;
      /* Search for number also in exponent */
      number_found = FALSE;

      /* Optional '+' or '-' in exponent */
      if (*cp == '-' || *cp == '+')
        cp++;

      /* Exponent number */
      while (*cp >= '0' && *cp <= '9')
      {
        cp++;
        number_found = TRUE;
      }
    }

    /* Length check - len <= 0 means do not check length */
    if (len > 0 && (cp - string) > len)
        return FALSE;

    /* Update *end only when number is found */
    if (number_found) {
      *end = cp;
    }

    return number_found;
}
```

Reply:
Thanks! Now I fixed whole detection of numbers.